src/game/UIExitMenu.cpp

```diff
@@ -0,0 +1,180 @@
+                                30001, FX_BUNDLE_MOVE_DIR_FLEXABLETARGET);
+        v0 = s_pPlayer;
+    }
+    
+    */
+    // TODO: CN3FXMgr::TriggerCursor
+    CGameBase::s_pPlayer->StunRelease();
+
+    if (CGameProcedure::s_pProcMain->m_pSubProcPerTrade) {
+
+        // TODO: MERCHANT CLEAR MODEL
+
+        CGameProcedure::s_pProcMain->m_pSubProcPerTrade->LeavePerTradeState(PER_TRADE_RESULT_MY_CANCEL);
+    }
+
+    if (CGameProcedure::s_pProcMain->m_pUIPartyOrForce) {
+        CGameProcedure::s_pProcMain->m_pUIPartyOrForce->MemberDestroy();
+    }
+
+    std::string szIP = CGameProcedure::s_pSocket->GetCurrentIP();
+    DWORD       dwPort = CGameProcedure::s_pSocket->GetCurrentPort();
+
+    CGameProcedure::s_bNeedReportConnectionClosed = false;
+    CGameProcedure::s_pSocket->Disconnect();
+    Sleep(4000);
+    CGameProcedure::s_pSocket->Connect(CGameProcedure::s_hWndBase, szIP.c_str(), dwPort);
+    CGameProcedure::s_bNeedReportConnectionClosed = true;
+
+    CGameProcedure::MsgSend_GameServerLogIn();
+    CGameProcedure::ProcActiveSet((CGameProcedure *)CGameProcedure::s_pProcCharacterSelect);
+}
+
+void CUIExitMenu::AddWarningMessage(DWORD idMessage, bool visbleMenu) {
+
+    if (!CGameProcedure::s_pProcMain->m_pUIChatDlg) {
+        return;
+    }
+
+    std::string szMsg;
+    ::_LoadStringFromResource(idMessage, szMsg);
+
+    CGameProcedure::s_pProcMain->m_pUIChatDlg->AddChatMsg(N3_CHAT_NORMAL, szMsg, 0xFFFF0000);
+
+    SetVisible(visbleMenu);
+}
+
+bool CUIExitMenu::ReceiveMessage(CN3UIBase * pSender, DWORD dwMsg) {
+    if (NULL == pSender) {
+        return false;
+    }
+
+    if (dwMsg == UIMSG_BUTTON_CLICK) {
+
+        if (pSender->m_szID == "btn_chr") { // Select characeter
+
+            if (CGameProcedure::s_pProcMain->m_pExitState == EXIT_STATE_ALLOW_LEAVE) {
+                SelectCharacter();
+            }
+
+            AddWarningMessage(IDS_EXIT_GAME_DURING_BATTLE_WARNING, false);
+            CGameProcedure::s_pProcMain->m_pExitType = EXIT_TYPE_SELECTCHAR;
+        }
+
+        else if (pSender->m_szID == "btn_exit") {
+            if (CGameProcedure::s_pProcMain->m_pExitState == EXIT_STATE_ALLOW_LEAVE) {
+                ::PostQuitMessage(0);
+            }
+
+            AddWarningMessage(IDS_EXIT_GAME_DURING_BATTLE_WARNING, false);
+
+            CGameProcedure::s_pProcMain->m_pExitType = EXIT_TYPE_EXIT;
+        } else if (pSender->m_szID == "btn_option") {
+            if (CGameProcedure::s_pProcMain->m_pExitState == EXIT_STATE_ALLOW_LEAVE) {
+                ::ShellExecute(NULL, "open", "Option.exe", NULL, NULL, SW_SHOWNORMAL);
+                PostQuitMessage(0);
+            }
+
+            AddWarningMessage(IDS_EXIT_GAME_DURING_BATTLE_WARNING, false);
+            CGameProcedure::s_pProcMain->m_pExitType = EXIT_TYPE_OPTION;
+        } else if (pSender->m_szID == "btn_cancel") {
+            SetVisible(false);
+        }
+    }
+    return true;
```

Comment:
I would refactor this entire function to this:
```c++

    if (dwMsg == UIMSG_BUTTON_CLICK) {
        if (pSender == (CN3UIBase *)m_pBtnChr) {
            if (CGameProcedure::s_pProcMain->m_pExitState == EXIT_STATE_ALLOW_LEAVE) {
                //CGameProcedure::s_pProcMain->m_bIsSeekingParty = false; // TODO: Reset seeking party.
                SelectCharacter();
            } else {
                AddWarningMessage(IDS_EXIT_GAME_DURING_BATTLE_WARNING, false);
                CGameProcedure::s_pProcMain->m_pExitType = EXIT_TYPE_SELECTCHAR;
            }
        } else if (pSender == (CN3UIBase *)m_pBtnExit) {
            if (CGameProcedure::s_pProcMain->m_pExitState == EXIT_STATE_ALLOW_LEAVE) {
                ::PostQuitMessage(0);
            } else {
                AddWarningMessage(IDS_EXIT_GAME_DURING_BATTLE_WARNING, false);
                CGameProcedure::s_pProcMain->m_pExitType = EXIT_TYPE_EXIT;
            }
        } else if (pSender == (CN3UIBase *)m_pBtnOption) {
            if (CGameProcedure::s_pProcMain->m_pExitState == EXIT_STATE_ALLOW_LEAVE) {
                ::ShellExecute(NULL, "open", "Option.exe", NULL, NULL, SW_SHOWNORMAL);
                PostQuitMessage(0);
            } else {
                AddWarningMessage(IDS_EXIT_GAME_DURING_BATTLE_WARNING, false);
                CGameProcedure::s_pProcMain->m_pExitType = EXIT_TYPE_OPTION;
            }
        } else if (pSender == (CN3UIBase *)m_pBtnCancel) {
            SetVisible(false);
        }
    }

    return true;
```

Reply:
Fixed here: https://github.com/ko4life-net/ko/pull/166/commits/6799f3d73452fc49bf88a81a5d73ad7df3e55e59